src/libraries/Microsoft.Extensions.Hosting/src/HostApplicationBuilder.cs

```diff
@@ -0,0 +1,435 @@
+        /// and so the delegate provided by <paramref name="configure"/> will run after all other services have been registered.
+        /// </para>
+        /// <para>
+        /// Multiple calls to <see cref="ConfigureContainer{TContainerBuilder}(IServiceProviderFactory{TContainerBuilder}, Action{TContainerBuilder})"/> will replace
+        /// the previously stored <paramref name="factory"/> and <paramref name="configure"/> delegate.
+        /// </para>
+        /// </remarks>
+        public void ConfigureContainer<TContainerBuilder>(IServiceProviderFactory<TContainerBuilder> factory, Action<TContainerBuilder>? configure = null) where TContainerBuilder : notnull
+        {
+            _createServiceProvider = () =>
+            {
+                TContainerBuilder containterBuilder = factory.CreateBuilder(Services);
+                // Call _configureContainer in case anyone adds more callbacks via HostBuilderAdapter.ConfigureContainer<TContainerBuilder>() during build.
+                // Otherwise, this is equivalent to configure?.Invoke(containerBuilder).
+                _configureContainer(containterBuilder);
+                return factory.CreateServiceProvider(containterBuilder);
+            };
+
+            // Store _configureContainer separately so it can replaced individually by the HostBuilderAdapter.
+            _configureContainer = containerBuilder => configure?.Invoke((TContainerBuilder)containerBuilder);
+        }
+
+        /// <summary>
+        /// Build the host. This can only be called once.
+        /// </summary>
+        /// <returns>An initialized <see cref="IHost"/>.</returns>
+        public IHost Build()
+        {
+            if (_hostBuilt)
+            {
+                throw new InvalidOperationException(SR.BuildCalled);
+            }
+            _hostBuilt = true;
+
+            var hostBuilderAdapter = new HostBuilderAdapter(this);
+            using DiagnosticListener diagnosticListener = HostBuilder.LogHostBuilding(hostBuilderAdapter);
+            hostBuilderAdapter.ApplyChanges();
+
+            _appServices = _createServiceProvider();
+
+            // Prevent further modification of the service collection now that the provider is built.
+            _checkedServiceCollection.IsReadOnly = true;
+
+            return HostBuilder.ResolveHost(_appServices, diagnosticListener);
+        }
+
+        private sealed class HostBuilderAdapter : IHostBuilder
+        {
+            private readonly HostApplicationBuilder _hostApplicationBuilder;
+
+            private readonly List<Action<IConfigurationBuilder>> _configureHostConfigActions = new();
+            private readonly List<Action<HostBuilderContext, IConfigurationBuilder>> _configureAppConfigActions = new();
+            private readonly List<IConfigureContainerAdapter> _configureContainerActions = new();
+            private readonly List<Action<HostBuilderContext, IServiceCollection>> _configureServicesActions = new();
```

Comment:
Have we thought about not allocating all these lists, or even this HostBuilderAdapter at all, when no one is listening to the LogHostBuilding event?